packages/wdio-browserstack-service/src/instrumentation/performance/performance-tester.ts

```diff
@@ -0,0 +1,329 @@
+        })
+        const dat = Object.entries(times).map(([key, value]) => {
+            return {
+                name: key,
+                time: value
+            }
+        })
+        this._csvWriter.writeRecords(dat)
+            .then(() => BStackLogger.info('Performance CSV report generated successfully'))
+            .catch((error: any) => console.error(error))
+    }
+
+    static Measure(label: string, details: PerformanceDetails = {}) {
+        const self = this
+        return (
+            target: Object,
+            key: string | symbol,
+            descriptor: TypedPropertyDescriptor<any>) => {
+            const originalMethod: Function = descriptor.value
+            if (descriptor.value) {
+                descriptor.value = function() {
+                    return PerformanceTester.measure.apply(self, [label, originalMethod as Function, { methodName: key.toString(), ...details }, arguments, this])
+                }
+            }
+        }
+    }
+
+    static measureWrapper(name: string, fn: Function, details: PerformanceDetails = {}) {
+        const self = this
+
+        details.worker = PerformanceTester.getProcessId()
+        details.testName = PerformanceTester.scenarioThatRan?.pop()
+        details.platform = PerformanceTester.browser?.sessionId
+
+        return function (...args: any[]) {
+            const methodArgs = [name, fn, details, args]
+
+            return self.measure.apply(self, methodArgs as [string, Function, {}, IArguments?])
+        }
+
+    }
+
+    static isEnabled() {
+        return !(process.env.BROWSERSTACK_SDK_INSTRUMENTATION === 'false')
+    }
+
+    static measure(label: string, fn: Function, details = {}, args?: IArguments, thisArg: any = null) {
+        try {
+            if (this.started && this.isEnabled()) {
+                PerformanceTester.start(label)
+                this.details && (this.details[label] = details)
+
+                try {
+                    const returnVal = fn.apply(thisArg, args)
+                    if (returnVal instanceof Promise) {
+                        return new Promise((resolve, reject) => {
+                            returnVal
+                                .then(v => {
+                                    PerformanceTester.end(label)
+                                    resolve(v)
+                                }).catch(e => {
+                                    PerformanceTester.end(label, false, util.format(e))
+                                    reject(e)
+                                })
+                        })
+                    }
+
+                    PerformanceTester.end(label)
+
+                    return returnVal
+                } catch (er) {
+                    PerformanceTester.end(label, false, util.format(er))
+                    throw er
+                }
+            }
+
+            return fn.apply(thisArg, args)
+        } catch (err) {
+            return fn.apply(thisArg, args)
+        }
```

Comment:
Comments:

- don't wrap everything into try catch
- no need to wrap a promise into another promise
- if you use nested try/catch you almost always do something wrong (complexity)
- `PerformanceTester.end(label)` was called twice

```suggestion
        if (!this.started || !this.isEnabled()) {
            return
        }

        try {
            if (this.started && this.isEnabled()) {
                PerformanceTester.start(label)
                this.details && (this.details[label] = details)

                try {
                    const returnVal = fn.apply(thisArg, args)
                    if (returnVal instanceof Promise) {
                        await returnVal
                    }

                    PerformanceTester.end(label)
                    return returnVal
                } catch (er) {
                    PerformanceTester.end(label, false, util.format(er))
                    throw er
                }
            }

            return fn.apply(thisArg, args)
        } catch (err) {
            return fn.apply(thisArg, args)
        }
```

Reply:
removed unnecessary try..catch.
If we use `await`,  have to make this function as async. Same for the measureWrapper where this is used. So, we have to use always await while using measureWrapper. That's why returning a Promise and using Performance.end